<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/assignment/ass6/hpdp/ObehSong/Assignment_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Assignment 6: Mastering Big Data Handling <a href="https://colab.research.google.com/drive/15hRsu4BFq0sPPOanMgzvImAaHCHYJ5y_?usp=sharing"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Group Name: ObehSong**
Group Members

| Name                                     | Matrix Number
| :---------------------------------------- | :-------------:
| FONG KHAH KHEH | A21EC0026
| KEE SHIN PEARL | A21EC0190

# About Dataset

##**eCommerce behavior data from multi category store**

This dataset contains 285 million users' events from eCommerce website. This dataset contains behavior data spanning seven months, from October 2019 to April 2020, collected from a large multi-category online store. The data is part of the Open CDP project and captures various events related to user interactions with products on the platform.


**File structure**


| Property         | Description                                               |
|------------------|-----------------------------------------------------------|
| event_time       | Time when the event occurred (in UTC).                    |
| event_type       | Type of event, including view, cart, remove_from_cart, and purchase. |
| product_id       | ID of the product.                                        |
| category_id      | Product's category ID.                                    |
| category_code    | Product's category taxonomy (code name).                  |
| brand            | Downcased string of the brand name (optional, can be missing). |
| price            | Float price of the product.                               |
| user_id          | Permanent user ID.                                        |
| user_session     | Temporary user's session ID, changing with each return to the online store. |






# Downloading the Dataset

Please use the link below to access the dataset:

[eCommerce behavior data from multi category store](https://www.kaggle.com/datasets/mkechinov/ecommerce-behavior-data-from-multi-category-store)


##Upload Kaggle API Token

In [1]:
from google.colab import files

# Upload kaggle.json
uploaded = files.upload()

Saving kaggle (1).json to kaggle (1).json


## Download and Unzip Dataset

In [2]:
# Move the uploaded kaggle (1).json file to the required directory
!mkdir -p ~/.kaggle
!mv '/content/kaggle (1).json' ~/.kaggle/kaggle.json
!chmod 600 ~/.kaggle/kaggle.json

# Download and unzip the dataset using the Kaggle API
!kaggle datasets download -d mkechinov/ecommerce-behavior-data-from-multi-category-store

# Unzip the downloaded dataset
!unzip -q ecommerce-behavior-data-from-multi-category-store.zip

ecommerce-behavior-data-from-multi-category-store.zip: Skipping, found more recently modified local copy (use --force to force download)
replace 2019-Nov.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
replace 2019-Oct.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


The dataset has been downloaded and extracted.

#Load the full dataset

importing libraries

In [4]:
import numpy as np
import pandas as pd

Read the dataset

In [4]:
df = pd.read_csv('2019-Oct.csv')
df.head()

event_time event_type  product_id          category_id  \
0  2019-10-01 00:00:00 UTC       view    44600062  2103807459595387724   
1  2019-10-01 00:00:00 UTC       view     3900821  2053013552326770905   
2  2019-10-01 00:00:01 UTC       view    17200506  2053013559792632471   
3  2019-10-01 00:00:01 UTC       view     1307067  2053013558920217191   
4  2019-10-01 00:00:04 UTC       view     1004237  2053013555631882655   

                         category_code     brand    price    user_id  \
0                                  NaN  shiseido    35.79  541312140   
1  appliances.environment.water_heater      aqua    33.20  554748717   
2           furniture.living_room.sofa       NaN   543.10  519107250   
3                   computers.notebook    lenovo   251.74  550050854   
4               electronics.smartphone     apple  1081.98  535871217   

                           user_session  
0  72d76fde-8bb3-4e00-8c23-a032dfed738c  
1  9333dfbd-b87a-4708-9857-6336556b0fcc  
2  566511c2-e2e3-422b-b695-cf8e6e792ca8  
3  7c90fc70-0e80-4590-96f3-13c02c18c713  
4  c6bd7419-2748-4c56-95b4-8cec9ff8b80d

Memory usage of the original dataset:

In [5]:
print("Memory usage:", df.memory_usage().sum() / (1024 ** 2), "MB")

Memory usage: 2914.7254333496094 MB


Computation time of the original dataset:

In [7]:
import timeit
time = timeit.timeit(lambda:df, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.0510000265639974e-06 seconds


File size of the original dataset:

In [8]:
# Calculate and print the estimated file size of the original dataset
original_file_size = df.memory_usage(deep=True).sum()
print("File Size of the Original Dataset:", original_file_size / (1024 ** 2), "MB")

File Size of the Original Dataset: 15743.616793632507 MB


# Strategies for Big Datasets

## Load Less Data : Strategically load only the essential portions of the dataset to optimize memory usage.

We want to know the user behavior and purchase details. Therefore, we select the columns 'event_time,' 'event_type,' 'product_id,' 'price,' 'user_id,' and 'user_session.'

In [5]:
import pandas as pd

# Specify the columns you want to load
columns_to_load =  ['event_time', 'event_type', 'product_id', 'price', 'user_id', 'user_session']

# Load the dataset with selected columns
df_lessdata = pd.read_csv('2019-Oct.csv', usecols=columns_to_load)

In [6]:
# Display the loaded DataFrame
df_lessdata.head()

event_time event_type  product_id    price    user_id  \
0  2019-10-01 00:00:00 UTC       view    44600062    35.79  541312140   
1  2019-10-01 00:00:00 UTC       view     3900821    33.20  554748717   
2  2019-10-01 00:00:01 UTC       view    17200506   543.10  519107250   
3  2019-10-01 00:00:01 UTC       view     1307067   251.74  550050854   
4  2019-10-01 00:00:04 UTC       view     1004237  1081.98  535871217   

                           user_session  
0  72d76fde-8bb3-4e00-8c23-a032dfed738c  
1  9333dfbd-b87a-4708-9857-6336556b0fcc  
2  566511c2-e2e3-422b-b695-cf8e6e792ca8  
3  7c90fc70-0e80-4590-96f3-13c02c18c713  
4  c6bd7419-2748-4c56-95b4-8cec9ff8b80d

In [7]:
df_lessdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42448764 entries, 0 to 42448763
Data columns (total 6 columns):
 #   Column        Dtype  
---  ------        -----  
 0   event_time    object 
 1   event_type    object 
 2   product_id    int64  
 3   price         float64
 4   user_id       int64  
 5   user_session  object 
dtypes: float64(1), int64(2), object(3)
memory usage: 1.9+ GB


Memory usage of the df_lessdata

In [8]:
print("Memory usage:", df_lessdata.memory_usage().sum() / (1024 ** 2), "MB")

Memory usage: 1943.1503295898438 MB


Computation time of the df_lessdata

In [9]:
import timeit
time = timeit.timeit(lambda:df_lessdata, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.742999984344351e-06 seconds


File size of the df_lessdata

In [18]:
# Calculate and print the estimated file size of the df_lessdata
df_lessdata_file_size = df_lessdata.memory_usage(deep=True).sum()
print("File Size of the df_lessdata Dataset:", df_lessdata_file_size / (1024 ** 2), "MB")

File Size of the df_lessdata Dataset: 10447.266115188599 MB


## Use Chunking : Process the data in smaller pieces to avoid memory issues.

 Sets the variable chunksize to 20,000. This value represents the number of rows that will be read into each chunk. In other words, the dataset will be processed in chunks, and each chunk will contain 20,000 rows of data.

In [11]:
chunksize = 20000
chunk = pd.read_csv('2019-Oct.csv', chunksize=20000)

In [12]:
for chunks in chunk:
    print(chunks.head())
    break

                event_time event_type  product_id          category_id  \
0  2019-10-01 00:00:00 UTC       view    44600062  2103807459595387724   
1  2019-10-01 00:00:00 UTC       view     3900821  2053013552326770905   
2  2019-10-01 00:00:01 UTC       view    17200506  2053013559792632471   
3  2019-10-01 00:00:01 UTC       view     1307067  2053013558920217191   
4  2019-10-01 00:00:04 UTC       view     1004237  2053013555631882655   

                         category_code     brand    price    user_id  \
0                                  NaN  shiseido    35.79  541312140   
1  appliances.environment.water_heater      aqua    33.20  554748717   
2           furniture.living_room.sofa       NaN   543.10  519107250   
3                   computers.notebook    lenovo   251.74  550050854   
4               electronics.smartphone     apple  1081.98  535871217   

                           user_session  
0  72d76fde-8bb3-4e00-8c23-a032dfed738c  
1  9333dfbd-b87a-4708-9857-6336556b0fc

Memory usage of the chunks

In [15]:
print("Memory usage:", chunks.memory_usage().sum() / (1024 ** 2), "MB")

Memory usage: 1.3734130859375 MB


Computation time of the chunks

In [16]:
import timeit
time = timeit.timeit(lambda:chunks, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.3910000689065782e-06 seconds


File size of the chunks

In [19]:
# Calculate and print the estimated file size of the dataset
chunks_file_size = chunks.memory_usage(deep=True).sum()
print("File Size of the Chunks Dataset:", chunks_file_size / (1024 ** 2), "MB")

File Size of the Chunks Dataset: 7.413230895996094 MB


## Optimize Data Types : Fine-tune data types to maximize efficiency and minimize memory consumption.

Optimizing data types is a strategy to reduce memory usage by choosing the most memory-efficient data types for each column in dataset.


When you load the dataset into pandas dataframe, the default datatypes assigned to each column are not memory efficient.

If we can convert these data types into memory-efficient ones we can save a lot of memory. For example, int64 can be downcast to int8 or int16 or int32 depending upon the max and min value the column holds.The below code takes of this downcasting for all numeric datatypes excluding object and DateTime types.

In [3]:
import pandas as pd

#Load the Data
df_optimize = pd.read_csv('2019-Oct.csv')

In [7]:
def reduce_mem_usage(df, verbose=True):
    # Define the numeric data types to consider for optimization
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']

    # Calculate the initial memory usage of the DataFrame
    start_mem = df.memory_usage().sum() / 1024**2

    # Iterate over each column in the DataFrame
    for col in df.columns:
        # Get the data type of the column
        col_type = df[col].dtypes

        # Check if the data type is numeric
        if col_type in numerics:
            # Calculate the minimum and maximum values of the column
            c_min = df[col].min()
            c_max = df[col].max()

            # Convert integer columns to the smallest possible integer type
            if str(col_type)[:3] == 'int':
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)
            # Convert float columns to the smallest possible float type
            else:
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)

    # Calculate the memory usage after optimization
    end_mem = df.memory_usage().sum() / 1024**2

    # Print the optimization results
    if verbose:
        print('Memory usage after optimization is: {:.2f} MB'.format(end_mem))
        print('Decreased by {:.1f}%'.format(100 * (start_mem - end_mem) / start_mem))

    return df


In [9]:
import pandas as pd
import numpy as np
df_opt_data=reduce_mem_usage(df_optimize)

Memory usage after optimization is: 2347.97 MB
Decreased by 19.4%


In [10]:
df_opt_data.head()

event_time event_type  product_id          category_id  \
0  2019-10-01 00:00:00 UTC       view    44600062  2103807459595387724   
1  2019-10-01 00:00:00 UTC       view     3900821  2053013552326770905   
2  2019-10-01 00:00:01 UTC       view    17200506  2053013559792632471   
3  2019-10-01 00:00:01 UTC       view     1307067  2053013558920217191   
4  2019-10-01 00:00:04 UTC       view     1004237  2053013555631882655   

                         category_code     brand       price    user_id  \
0                                  NaN  shiseido    35.78125  541312140   
1  appliances.environment.water_heater      aqua    33.18750  554748717   
2           furniture.living_room.sofa       NaN   543.00000  519107250   
3                   computers.notebook    lenovo   251.75000  550050854   
4               electronics.smartphone     apple  1082.00000  535871217   

                           user_session  
0  72d76fde-8bb3-4e00-8c23-a032dfed738c  
1  9333dfbd-b87a-4708-9857-6336556b0fcc  
2  566511c2-e2e3-422b-b695-cf8e6e792ca8  
3  7c90fc70-0e80-4590-96f3-13c02c18c713  
4  c6bd7419-2748-4c56-95b4-8cec9ff8b80d

In [13]:
df_opt_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42448764 entries, 0 to 42448763
Data columns (total 9 columns):
 #   Column         Dtype  
---  ------         -----  
 0   event_time     object 
 1   event_type     object 
 2   product_id     int32  
 3   category_id    int64  
 4   category_code  object 
 5   brand          object 
 6   price          float16
 7   user_id        int32  
 8   user_session   object 
dtypes: float16(1), int32(2), int64(1), object(5)
memory usage: 2.3+ GB


Memory usage of the df_opt_data

In [14]:
print("Memory usage:", df_opt_data.memory_usage().sum() / (1024 ** 2), "MB")

Memory usage: 2347.973289489746 MB


Computation time of the df_opt_data

In [15]:
import timeit
time = timeit.timeit(lambda:df_opt_data, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.27099974633893e-06 seconds


File size of the df_opt_data

In [17]:
df_opt_data_file_size = df_opt_data.memory_usage(deep=True).sum()
print("File Size of the df_opt_data Dataset:", df_opt_data_file_size / (1024 ** 2), "MB")

File Size of the df_opt_data Dataset: 15176.864649772644 MB


##Sampling : Implement sampling methodologies to extract meaningful insights from a subset of the dataset.

 The variable n_samples is defined to specify the desired number of samples, representing the size of the subset. In this case, 20,000 samples are chosen for analysis.

In [19]:
# Specify the number of samples you want
n_samples = 20000  # Adjust the number as needed

# Perform random sampling
df_sampled = df_optimize.sample(n=n_samples, random_state=42)  # Set a fixed random state for reproducibility

In [20]:
# Display the sampled DataFrame
df_sampled.head()

event_time event_type  product_id          category_id  \
10257365  2019-10-09 02:46:19 UTC       view    29900078  2059484601444729123   
28875143  2019-10-21 16:17:11 UTC       view     1801805  2053013554415534427   
28709595  2019-10-21 14:42:20 UTC       cart     1004870  2053013555631882655   
3458568   2019-10-03 18:40:22 UTC       view     5701087  2053013553970938175   
25352815  2019-10-19 08:58:51 UTC       view     5300097  2053013563173241677   

                    category_code      brand       price    user_id  \
10257365                      NaN       peda  1647.00000  532353313   
28875143     electronics.video.tv      artel   231.62500  536042235   
28709595   electronics.smartphone    samsung   284.00000  517143806   
3458568   auto.accessories.player        jvc    51.46875  556492770   
25352815                      NaN  remington    39.87500  561838859   

                                  user_session  
10257365  47059489-3a63-4353-9eab-0e239d5cb5aa  
28875143  a65ee34b-7701-4b25-95e7-45a1259e747e  
28709595  a9ba13a7-afac-41ca-a3bc-016c580908c3  
3458568   799d5c0a-e1f3-4f22-b61d-ab78cfed068c  
25352815  dd474f7f-0559-41f5-8be0-9d0912cf2b1b

Memory usage of the df_sampled

In [21]:
print("Memory usage:", df_sampled.memory_usage().sum() / (1024 ** 2), "MB")

Memory usage: 1.25885009765625 MB


Computation time of the df_sampled

In [22]:
import timeit
time = timeit.timeit(lambda:df_sampled, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.9180001800123136e-06 seconds


File size of the df_sampled

In [23]:
df_sampled_file_size = df_sampled.memory_usage(deep=True).sum()
print("File Size of the df_sampled Dataset:", df_sampled_file_size / (1024 ** 2), "MB")

File Size of the df_sampled Dataset: 7.29876708984375 MB


## Parallelize with Dask

In [64]:
!pip install dask

 imports the Dask DataFrame module

In [65]:
# Load the Pkgs
import dask.dataframe as dd

Read CSV with Dask

In [66]:
df_dask = dd.read_csv('2019-Oct.csv')

In [68]:
df_dask.info()

<class 'dask.dataframe.core.DataFrame'>
Columns: 9 entries, event_time to user_session
dtypes: object(5), float64(1), int64(3)

In [69]:
# Describe
df_dask.describe()

Dask DataFrame Structure:
              product_id category_id    price  user_id
npartitions=1                                         
                 float64     float64  float64  float64
                     ...         ...      ...      ...
Dask Name: describe-numeric, 35 graph layers

In [30]:
# Specify the columns you want to load
columns_to_load = ['event_time', 'event_type', 'category_code', 'brand', 'price', 'user_id']

# Select the specified columns
df_dask_selected = df_dask[columns_to_load]

# Compute the mean of the 'price' column
price_mean = df_dask_selected['price'].mean()

# Compute the result
result = price_mean.compute()

# Display the result
print(result)

290.32366068491416


Memory usage of the df_dask

In [52]:
print("Memory usage:",  df_dask.memory_usage(deep=True).sum().compute() / (1024 ** 2), "MB")

Memory usage: 15743.627413749695 MB


In [71]:
# Print the memory usage
print("Memory usage:", df_dask.memory_usage(deep=True).sum() / (1024 ** 2), "MB")

Memory usage: dd.Scalar<truediv..., dtype=float64> MB


Computation time of the df_dask

In [53]:
import timeit
time = timeit.timeit(lambda:df_dask, number=1)
print("Computation Time:", time, "seconds")

Computation Time: 1.8910004655481316e-06 seconds


File size of the df_dask

In [72]:
df_dask_file_size = df_dask.memory_usage(deep=True).sum().compute()
print(f"Total file size of the Dask DataFrame:", df_dask_file_size / (1024 ** 2), "MB")

Total file size of the Dask DataFrame: 15743.627413749695 MB


#Comparative Analysis

|                | Traditional Methods               | Load Less Data                    | Use Chunking                      | Optimize Data Types               | Sampling                          | Parallelize with Dask             |
| -------------- | --------------------------------- | ---------------------------------- | --------------------------------- | ---------------------------------- | --------------------------------- | ---------------------------------- |
| Memory Usage   | 2914.73 MB                        | 1943.15 MB                        | 1.37 MB                           | 2347.97 MB                        | 1.26 MB                           | 15743.63 MB                       |
| Computation Time| 1.05e-06 seconds                   | 1.74e-06 seconds                   | 1.39e-06 seconds                   | 1.27e-06 seconds                   | 1.92e-06 seconds                   | 1.46e-06 seconds                   |
| File Size      | 15743.62 MB                       | 10447.27 MB                       | 7.41 MB                           | 15176.86 MB                       | 7.30 MB                           | 15743.63 MB                       |


**Traditional Methods:**

Memory Usage: High memory usage (2914.73 MB) as the entire dataset is loaded into memory.
Computation Time: Very fast computation time (1.05e-06 seconds), but memory-intensive.
File Size (on Disk): High file size (15743.62 MB) as the full dataset is stored.

Load Less Data:
**bold text**
Memory Usage: Optimized memory usage (1943.15 MB) as only essential portions of the dataset are loaded.
Computation Time: Fast computation time (1.74e-06 seconds) due to loading only necessary data.
File Size (on Disk): Lower file size (10447.27 MB) compared to traditional methods.

**Use Chunking:**

Memory Usage: Highly optimized memory usage (1.37 MB) by processing data in smaller, manageable chunks.
Computation Time: Fast computation time (1.39e-06 seconds) due to parallel processing of smaller chunks.
File Size (on Disk): Significantly reduced file size (7.41 MB) compared to full dataset.

**Optimize Data Types:**

Memory Usage: Optimized memory usage (2347.97 MB) by fine-tuning data types.
Computation Time: Fast computation time (1.27e-06 seconds) due to efficient data types.
File Size (on Disk): Similar file size (15176.86 MB) to traditional methods, but with memory efficiency.

**Sampling:**

Memory Usage: Highly reduced memory usage (1.26 MB) by working with a sample of the data.
Computation Time: Fast computation time (1.92e-06 seconds) on the sampled data.
File Size (on Disk): Significantly reduced file size (7.30 MB) compared to full dataset.

**Parallelize with Dask:**

Memory Usage: High memory usage (15743.63 MB) due to Dask's distributed computing handling large chunks of data.
Computation Time: Very fast computation time (1.46e-06 seconds) thanks to parallel and distributed computing.
File Size (on Disk): Total file size (15743.63 MB) represents the distributed storage used by Dask.

**Insights:**

Memory Usage: Chunking and sampling significantly reduce memory usage, making them suitable for large datasets.
Computation Time: Dask provides optimal computation time through parallel and distributed computing.
File Size (on Disk): Chunking and sampling result in substantially smaller file sizes, making them efficient for storage.


# **Summary**

Traditional Methods:
- Memory Usage: High (Entire dataset loaded)
- Computation Time: May be slower due to processing entire dataset
- File Size (on Disk): High (Entire dataset loaded)

Load Less Data:
- Memory Usage: Optimized (Loads only essential portions)
- Computation Time: Faster (Loads only essential data)
- File Size (on Disk): Lower (Only loads necessary data)

Use Chunking:
- Memory Usage: Optimized (Processes data in smaller pieces)
- Computation Time: Faster (Processes smaller chunks of data)
- File Size (on Disk): Depends on Chunk Size (Reduced compared to full)

Optimize Data Types:
- Memory Usage: Optimized (Fine-tunes data types)
- Computation Time: Faster (Optimized data types)
- File Size (on Disk): Similar to Traditional (Optimization in-memory)

Sampling:
- Memory Usage: Reduced (Depends on Sample Size)
- Computation Time: Faster on Sampled Data
- File Size (on Disk): Depends on Sample Size (Reduced compared to full)

Parallelize with Dask:
- Memory Usage: Optimized (Dask operates on chunks)
- Computation Time: Optimal (Parallel and distributed computing)
- File Size (on Disk): Distributed Storage (Optimal for large datasets)
